lightning/src/ln/peers/handler.rs

```diff
@@ -0,0 +1,3280 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Top level peer message handling and socket handling logic lives here.
+//!
+//! Instead of actually servicing sockets ourselves we require that you implement the
+//! SocketDescriptor interface and use that to receive actions which you should perform on the
+//! socket, and call into PeerManager with bytes read from the socket. The PeerManager will then
+//! call into the provided message handlers (probably a ChannelManager and NetGraphmsgHandler) with messages
+//! they should handle, and encoding/sending response messages.
+
+use bitcoin::secp256k1::key::{SecretKey,PublicKey};
+
+use ln::features::InitFeatures;
+use ln::msgs;
+use ln::msgs::{ChannelMessageHandler, LightningError, RoutingMessageHandler};
+use ln::channelmanager::{SimpleArcChannelManager, SimpleRefChannelManager};
+use util::ser::{Writeable};
+use ln::wire;
+use ln::wire::{Encode, Message};
+use util::byte_utils;
+use util::events::{MessageSendEvent, MessageSendEventsProvider};
+use util::logger::Logger;
+use routing::network_graph::NetGraphMsgHandler;
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::{cmp,error,hash,fmt};
+use std::ops::Deref;
+
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::hashes::sha256::HashEngine as Sha256Engine;
+use bitcoin::hashes::{HashEngine, Hash};
+use ln::peers::outbound_queue::OutboundQueue;
+use ln::peers::transport::{PayloadQueuer, Transport};
+use bitcoin::hashes::core::iter::Filter;
+use std::collections::hash_map::IterMut;
+
+// Number of items that can exist in the OutboundQueue before Sync message flow control is triggered
+const OUTBOUND_QUEUE_SIZE: usize = 10;
+
+/// Interface PeerManager uses to interact with the Transport object
+pub(super) trait ITransport: MessageQueuer {
```

Comment:
I find it very hard with this design to see trivially where the queues are hiding, and how big they can grow, and what each trait means. Can we get some documentation on what the difference between a Message and PayloadQueuer is, how that fits into the OutboundQueue, why `process_input` takes a trait which `OutboundQueue` implements, etc. I also think we can squash a bunch of stuff to make this much more readable - can we avoid having both a PayloadQueuer and a MessageQueuer and only have one queue and one thing that encrypts/decrypts against the queue, or is that hiding here somewhere and I dont see it?

Reply:
Vec vs. Iterator

This was the design chosen in #494 and I chose to limit the number of behavior changes in the initial refactoring to ease the review (although I'm not sure how much that really saved in the end). I've done my best to separate out behavior changes to a separate patch so they can be easily dropped if necessary.

This could just as easily be a drain iterator or an iterator followed by `clear()` or `drain()`. I don't really like the pattern of iterators that also remove, so this was the cleanest analog. I don't think there is any additional memory pressure here since the bytes exist in the Decryptor anyway and they are effectively transferred to this layer.

But, it seems reasonable to change this up in the future if there is a valid reason. An iterator approach would mesh well with my thoughts in #698.

Reply:
I think cleaning up the naming might make this a bit easier. I'm not attached to any of the trait names but tried to follow Rust trait idioms, but I can see how they can all blend if you haven't been staring at the code for 6 weeks straight. I've answered your questions inline. If you have better names that make the traits a lot more obvious I am happy to clean them up.

> I find it very hard with this design to see trivially where the queues are hiding,

The only queue is the `OutboundQueue` and is owned by the `PeerManager` at the top level. Same as the LL before.

>  and how big they can grow, 

The `PeerManager` shouldn't care how big they can grow and this was one of the benefits of splitting out the `OutboundQueue` object. I've kept the const at the top level, but that was only to keep a single constructor for tests and the actual implementation.

If there are performance issues in the future, the features can just be added to the `OutboundQueue` and tested in isolation.

As far as max size, the `read_event()` path pushes unconditionally. There was no behavior change for this in my patch stack and the API and docs now reflect the behavior. Is this the best long-term, probably not. That feature is captured in #703.

>and what each trait means. Can we get some documentation on what the difference between a Message and PayloadQueuer is,

A naming fixup might help the confusion. I will do my best to describe the relationships here and your feedback can help me clean up the trait documentation to help the next reader.

The `PeerHandler` owns the `OutboundQueue` which is a simple wrapper object for raw bytes that need to be sent out of the socket descriptor. The `OutboundQueue` implements two traits: 

* `PayloadQueuer` handles the raw byte enqueue into the queue
* `SocketDescriptorFlusher` handles the flushing and flow control

The `PayloadQueuer` is the interface consumed by the `Transport` layer. It needs a way to push unencrypted bytes pre-NOISE and encrypted bytes post-NOISE.

> how that fits into the OutboundQueue, why `process_input` takes a trait which `OutboundQueue` implements, etc. 

`process_input()` takes a `PayloadQueuer` because it is just responsible for writing raw bytes into the `OutboundQueue`. By taking the trait, instead of the `OutboundQueue`, it is more clear (and guaranteed by the type system) that the `Transport` layer is only responsible for pushing items and not flushing them.

> I also think we can squash a bunch of stuff to make this much more readable 

Sort of opposite feedback from @ariard who asked me to split it out more. Is this something that can just be fixed with `git diff`? I did my best to not only break up the commits into smaller consumable pieces but also walk the reviewers through my train of thought as I did the refactor. Big refactors like this don't always have an obvious endpoint when you start. You have to untangle it piece by piece and see what it looks like. When I'm reviewing large changes like this I find it adds a lot more confidence if I can see how someone approached a redesign vs. just the final product.

In the end, if you can't review it this will just sit here forever, so please let me know how I can help make it easier on you.

> - can we avoid having both a PayloadQueuer and a MessageQueuer and only have one queue and one thing that encrypts/decrypts against the queue, or is that hiding here somewhere and I don't see it?

There is one queue. The raw byte (Payload) queue `OutboundQueue` is owned by the `PeerManager`.

The `PeerManager` only interacts with `MessageQueuer` since it only needs to send post-NOISE Messages

The `Transport` only interacts with `PayloadQueuer` since it just deals with raw bytes. The unencrypted ACT bytes and encrypted Messages.

During pre-NOISE, the `Transport` layer takes in a `PayloadQueuer` during `process_input()` where it will enqueue the unencrypted ACT bytes.

During post-NOISE, the `PeerManager` asks the `Transport` layer to encrypt bytes via `enqueue_message()` which is essentially: take this Message, encrypt it, and add it to the queue.

This separation allows the `PeerManager` to own the queue and give it to the `Transport` functions that need to enqueue onto it.

So, in the end, it matches with what you want... I think. There is one queue `OutboundQueue` and `Transport` encrypts/decrypts against it.

Reply:
Right, I think its possible with more documentation and trait renaming to make it clearer what's going on, maybe without ripping out *all* the test traits, though the ratio of traits-to-code seems difficult to swallow. eg it may make sense to reduce the trait count and take the complexity hit by adding more shims in the test side of the code.

Reply:
Done: `review: Collapse OutboundQueue/Transport traits`